### PR TITLE
Add force subsample argument to JPEG save

### PR DESCRIPTION
**Pull request by Elad-Laufer**
If Q >= 90 then subsampling is disabled, this argument forces it

### CODE REVIEW

**Comment by jcupitt** [comment]
Hello @Elad-Laufer,

We already have `no-subsample` and `Q`. Perhaps we should deprecate `no-subsample` and have a three way switch: force on, force off, automatic (set by Q), with auto as the default.

**Comment by lovell** [comment]
If there's a plan to offer more fine-grained control over subsampling, perhaps we should consider allowing 4:2:2 as well as 4:4:4 ("no-subsample") and 4:2:0 ("force-subsample")?

**Comment by jcupitt** [comment]
I looked into this when we added `jpeg-chroma-subsample` to jpegload: the only two supported settings for JPEG are 4:4:4 (no subsampling) and 4:2:0 (subsample both Cb and Cr by 2). It's very confusing :(

http://poynton.ca/PDFs/Chroma_subsampling_notation.pdf

**Comment by jcupitt** [comment]
(or that's my understanding from that useful PDF)

**Comment by lovell** [comment]
@jcupitt Thanks for the info, no worries, let's keep this as simple as possible.

File: libvips/foreign/jpegsave.c [hunk 1]
```diff
@@ -194,6 +198,13 @@ vips_foreign_save_jpeg_class_init( VipsForeignSaveJpegClass *class )
 		G_STRUCT_OFFSET( VipsForeignSaveJpeg, quant_table ),
 		0, 8, 0 );
 
+    VIPS_ARG_ENUM( class, "subsample_mode", 19,
+            _( "Subsample mode" ),
+            _( "Select chroma subsample operation mode" ),
+            VIPS_ARGUMENT_OPTIONAL_INPUT,
+            G_STRUCT_OFFSET( VipsForeignSaveJpeg, subsample_mode ),
+            VIPS_TYPE_FOREIGN_JPEG_SUBSAMPLE,
+            VIPS_FOREIGN_JPEG_SUBSAMPLE_AUTO );
```
Comment (jcupitt) [inline libvips/foreign/jpegsave.c]: I think we should tag `no_subsample` as deprecated. That'll hide it from APIs and docs, but it'll still work from old code.

File: libvips/foreign/jpegsave.c [hunk 2]
```diff
@@ -230,7 +241,7 @@ vips_foreign_save_jpeg_target_build( VipsObject *object )
 		jpeg->Q, jpeg->profile, jpeg->optimize_coding, 
 		jpeg->interlace, save->strip, jpeg->no_subsample,
```
Comment (jcupitt) [inline libvips/foreign/jpegsave.c]: Let's not pass no_subsample. Instead, a little higher up, have something like:

```C
if arg-set("no_subsample") && !arg_set("subsample_mode")
  subsample_mode = no_subsample ? OFF : ON
```

ie. old code using the `no_subsample` arg is able to set the default for the new `subsample_mode` value.

File: libvips/foreign/jpegsave.c [hunk 3]
```diff
@@ -297,7 +307,7 @@ vips_foreign_save_jpeg_file_build( VipsObject *object )
...
```
Comment (jcupitt) [inline libvips/foreign/jpegsave.c]: Again, remove `no_subsample`.

File: libvips/foreign/jpegsave.c [hunk 4]
```diff
@@ -367,7 +377,7 @@ vips_foreign_save_jpeg_buffer_build( VipsObject *object )
...
```
Comment (jcupitt) [inline libvips/foreign/jpegsave.c]: Remove 'no_subsample`.

File: libvips/foreign/jpegsave.c [hunk 5]
```diff
@@ -440,7 +450,7 @@ vips_foreign_save_jpeg_mime_build( VipsObject *object )
...
```
Comment (jcupitt) [inline libvips/foreign/jpegsave.c]: Remove.

File: libvips/foreign/pforeign.h [hunk 6]
```diff
@@ -166,7 +166,7 @@ int vips__jpeg_write_target( VipsImage *in, VipsTarget *target,
 	gboolean optimize_coding, gboolean progressive, gboolean strip,
 	gboolean no_subsample, gboolean trellis_quant,
```
Comment (jcupitt) [inline libvips/foreign/pforeign.h]: Oh heh and again.

File: libvips/foreign/vips2jpeg.c [hunk 7]
```diff
@@ -627,17 +628,24 @@ write_vips( Write *write, int qfac, const char *profile,
...
-	 */
-	if( no_subsample ||
-		qfac >= 90 ) { 
-		int i;
-
-		for( i = 0; i < in->Bands; i++ ) { 
-			write->cinfo.comp_info[i].h_samp_factor = 1;
-			write->cinfo.comp_info[i].v_samp_factor = 1;
-		}
+	switch (subsample_mode) {
```
Comment (jcupitt) [inline libvips/foreign/vips2jpeg.c]: libvips indentation style is K&R, so:

```C
switch( subsample_mode ) {
case VIPS_...ON:
        break;
```

ie. hard tabs set to 8, 80 char line length.

File: libvips/foreign/vips2jpeg.c [hunk 8]
```diff
@@ -627,17 +628,24 @@ write_vips( Write *write, int qfac, const char *profile,
...
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_ON:
+            break;
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_AUTO:
+            /* Turn off chroma subsampling. Follow IM and do it automatically for
+            * high Q.
```
Comment (jcupitt) [inline libvips/foreign/vips2jpeg.c]: Asterisks should line up.

```C
/* Like 
 * this.
 */
```

File: libvips/foreign/vips2jpeg.c [hunk 9]
```diff
@@ -627,17 +628,24 @@ write_vips( Write *write, int qfac, const char *profile,
...
+            */
+            if( qfac < 90 ) {
+                break;
+            }
+        case VIPS_FOREIGN_JPEG_SUBSAMPLE_OFF:
+        default: {
```
Comment (jcupitt) [inline libvips/foreign/vips2jpeg.c]: Newline before `{`, like a func definition, and `for` formatted as:

```C
for( i = 0; i < xxx; i++ ) { 
```

File: libvips/foreign/vips2jpeg.c [hunk 10]
```diff
@@ -797,12 +806,18 @@ vips__jpeg_write_target( VipsImage *in, VipsTarget *target,
 	 */
         target_dest( &write->cinfo, target );
 
+    /* Retain old behavior for now
+     */
+    if( no_subsample ) {
+        subsample_mode = VIPS_FOREIGN_JPEG_SUBSAMPLE_OFF;
+    }
+
```
Comment (jcupitt) [inline libvips/foreign/vips2jpeg.c]: This `if()` can go if we have the extra logic in `_build()`.

File: test/test-suite/test_foreign.py [hunk 11]
```diff
@@ -254,6 +254,27 @@ def jpeg_valid(im):
             # format area at the end
             assert y.startswith("hello world")
 
+    @skip_if_no("jpegload")
+    def test_jpegsave(self):
+        im = pyvips.Image.new_from_file(JPEG_FILE)
+
+        # higher Q should mean a bigger buffer
+        q10 = im.jpegsave_buffer(Q=10)
+        q10_subsample_off = im.jpegsave_buffer(Q=10, subsample_mode=2)
```
Comment (jcupitt) [inline test/test-suite/test_foreign.py]: I would use the string value here, ie. `"off"`.

File: test/test-suite/test_foreign.py [hunk 12]
```diff
@@ -254,6 +254,27 @@ def jpeg_valid(im):
...
+        q10_subsample_no = im.jpegsave_buffer(Q=10, no_subsample=1)
```
Comment (jcupitt) [inline test/test-suite/test_foreign.py]: I don't think we should have deprecated API being tested in the test suite, let's remove this.

Same for the case below. Just test `on`, `off`, `auto`.

File: libvips/foreign/vips2jpeg.c (hunk 7 above)
Comment (Elad-Laufer) [reply libvips/foreign/vips2jpeg.c]: done

File: libvips/foreign/vips2jpeg.c (hunk 8 above)
Comment (Elad-Laufer) [reply libvips/foreign/vips2jpeg.c]: done

File: libvips/foreign/vips2jpeg.c (hunk 9 above)
Comment (Elad-Laufer) [reply libvips/foreign/vips2jpeg.c]: done

File: test/test-suite/test_foreign.py (hunk 12 above)
Comment (Elad-Laufer) [reply test/test-suite/test_foreign.py]: done

File: test/test-suite/test_foreign.py (hunk 11 above)
Comment (Elad-Laufer) [reply test/test-suite/test_foreign.py]: done

File: libvips/foreign/jpegsave.c (hunk 1 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: done

File: libvips/foreign/jpegsave.c (hunk 2 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: Should I add the check to each `_build(...)` function? currently I only check once at `vips__jpeg_write_target(...)`.

Is there a way to perform the check in single point without code duplication?

Comment (jcupitt) [reply libvips/foreign/jpegsave.c]: Yes, add a build method to the base class for jpegsave and do the test after chaining up.

Comment (jcupitt) [reply libvips/foreign/jpegsave.c]: tiffsave has something a little like this:

https://github.com/libvips/libvips/blob/master/libvips/foreign/tiffsave.c#L153

A `_build` method on the base class and use of `isset` to set defaults.

File: libvips/foreign/vips2jpeg.c (hunk 10 above)
Comment (Elad-Laufer) [reply libvips/foreign/vips2jpeg.c]: done

File: libvips/foreign/jpegsave.c (hunk 5 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: done

File: libvips/foreign/jpegsave.c (hunk 4 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: done

File: libvips/foreign/jpegsave.c (hunk 3 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: done

File: libvips/foreign/jpegsave.c (hunk 2 above)
Comment (Elad-Laufer) [reply libvips/foreign/jpegsave.c]: done

File: libvips/foreign/pforeign.h (hunk 6 above)
Comment (Elad-Laufer) [reply libvips/foreign/pforeign.h]: done

**Comment by jcupitt** [comment]
I revised the formatting slightly and added some docs.

**Pull request merged**